src/sagas/task.saga.js

```diff
@@ -0,0 +1,339 @@
+          break;
+        case 'exit':
+          yield call(displayTaskComplete, task);
+          yield put(
+            completeTask(task, message.timestamp, message.wasSuccessful)
+          );
+          if (task.name === 'eject') {
+            yield put(loadDependencyInfoFromDisk(project.id, project.path));
+          }
+          break;
+      }
+    }
+  } finally {
+    /* child process has completed */
+  }
+}
+
+export function* taskAbort({ task }: { task: Task }): Saga<void> {
+  const { processId, name } = task;
+
+  yield call(killProcessId, processId);
+  yield call([ipcRenderer, ipcRenderer.send], 'removeProcessId', processId);
+
+  // Once the children are killed, we should dispatch a notification
+  // so that the terminal shows something about this update.
+  // My initial thought was that all tasks would have the same message,
+  // but given that we're treating `start` as its own special thing,
+  // I'm realizing that it should vary depending on the task type.
+  // TODO: Find a better place for this to live.
+  const abortMessage = isDevServerTask(name)
+    ? 'Server stopped'
+    : 'Task aborted';
+
+  yield put(receiveDataFromTaskExecution(task, chalk.bold.red(abortMessage)));
+}
+
+export function* displayTaskComplete(task: Task): Saga<void> {
+  // Send a message to add info to the terminal about the task being done.
+  // TODO: ASCII fish art?
+
+  const message = 'Task completed';
+
+  yield put(receiveDataFromTaskExecution(task, chalk.bold.green(message)));
+}
+
+export function* taskComplete({ task }: { task: Task }): Saga<void> {
+  if (task.processId) {
+    yield call(
+      [ipcRenderer, ipcRenderer.send],
+      'removeProcessId',
+      task.processId
+    );
+  }
+
+  // The `eject` task is special; after running it, its dependencies will
+  // have changed.
+  // TODO: We should really have a `EJECT_PROJECT_COMPLETE` action that does
+  // this instead.
+  if (task.name === 'eject') {
+    const project = yield select(getProjectById, task.projectId);
+
+    yield put(loadDependencyInfoFromDisk(project.id, project.path));
+  }
+}
+
+const createStdioChannel = (
+  child: any,
+  handlers: {
+    stdout: (emitter: any) => (data: string) => void,
+    stderr: (emitter: any) => (data: string) => void,
+    exit: (emitter: any) => (code: number) => void,
+  }
+) => {
+  return eventChannel(emitter => {
+    child.stdout.on('data', handlers.stdout(emitter));
+    child.stderr.on('data', handlers.stderr(emitter));
+    child.on('exit', handlers.exit(emitter));
+
+    return () => {
+      /* unsubscribe any listeners */
+    };
+
+    // use an expanding buffer because we don't want to lose any information
+    // passed up by the child process. initialize it at a length of 2 because
+    // at bare minimum we expect to have 2 messages queued at some point (as
+    // the exit channel completes, it should emit the return code of the process
+    // and then immediately END
+  }, buffers.expanding(2));
```

Comment:
Curious why we need to specify a buffer at all? If I'm reading the docs correctly, the default is to have an unlimited buffer size. It seems like `buffers.expanding` can also grow to infinite size, we're just setting an initial size?

It's very possible (likely, even) that I'm misunderstanding something though, so please let me know :)

Reply:
> Like actionChannel (Effect), eventChannel (a factory function, not an Effect) creates a Channel for events but from event sources other than the Redux Store.

> Note: messages on an eventChannel are not buffered by default. You have to provide a buffer to the eventChannel factory in order to specify buffering strategy for the channel (e.g. eventChannel(subscriber, buffer)). [See the API docs](https://github.com/redux-saga/redux-saga/tree/master/docs/api#buffers) for more info.

Since this isn't processing redux actions, I went with an `eventChannel`, which doesn't buffer by default. In this particular case we probably don't need a buffer (?) since the console output is handled immediately without any blocking asynchronous calls, but I wasn't confident of that, and should we ever implement some sort of async handlers for console output down the road then this pattern would be required.

I'll remove it since I don't believe it's necessary now and just leave in a comment about how it should be used in the future.

Reply:
ahh I see.

Yeah, I could go either way, but I'm tempted to ditch it for now since it isn't needed, and "buffers" is scary-sounding haha.